### PR TITLE
perform single-cell data splits

### PR DESCRIPTION
**Pull request by MikeLippincott**
This PR performs single-cell data splits for the sinlge-cell timelapse model:

This is quite the complex data splitting procedure.
The data is split into holdout data, training, validation, and testing.
The training and validation data only contains single-cells that have grond truth labels at the terminal time point.
While testing and holdout data contains celles that do and no not have ground truth labels at the terminal time point.

### CODE REVIEW

File: 7.sc_timelapse_model/scripts/0.data_cleaning.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+# In[7]:
+
+
+start_time = time.time()
+
+
+# In[8]:
+
+
+for well_fov in tqdm(list(dict_of_sc_well_fovs.keys()), desc="Processing Well-FOVs"):
```
Comment (gwaybio) [inline 7.sc_timelapse_model/scripts/0.data_cleaning.py:112]: Needs documentation - why is fuzzy matching necessary? Is this what the tracking is supposed to do? Unclear to me

File: 7.sc_timelapse_model/scripts/0.data_cleaning.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+
+# In[12]:
+
+
+# write the cleaned dataframe to a parquet file
+output_sc_file_path = pathlib.Path("../results/cleaned_sc_profile.parquet").resolve(
+    strict=False
+)
+output_sc_endpoint_file_path = pathlib.Path(
```
Comment (gwaybio) [inline 7.sc_timelapse_model/scripts/0.data_cleaning.py:184]: why save endpoint separately? Needs documentation

File: 7.sc_timelapse_model/scripts/1.data.splitting.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+if in_notebook:
+    from tqdm.notebook import tqdm
+else:
+    from tqdm import tqdm
+
+
+# In[2]:
+
+
+# write the cleaned dataframe to a parquet file
```
Comment (gwaybio) [inline 7.sc_timelapse_model/scripts/1.data.splitting.py:33]: copy+paste typo

File: 7.sc_timelapse_model/scripts/1.data.splitting.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+# split the data into 80, 10, 10 stratified by the well
+train_sc_w_ground_truth_df, test_sc_w_ground_truth_df = train_test_split(
+    cell_w_ground_truth_df,
+    test_size=0.2,
+    stratify=cell_w_ground_truth_df["Metadata_Well"],
+    random_state=0,
+)
+test_sc_w_ground_truth_df, val_sc_w_ground_truth_df = train_test_split(
+    test_sc_w_ground_truth_df,
+    test_size=0.5,
```
Comment (gwaybio) [inline 7.sc_timelapse_model/scripts/1.data.splitting.py:153]: nice, I see what you did here - probably worth adding a comment describing why this is 0.5

File: 7.sc_timelapse_model/scripts/1.data.splitting.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+print(f"test_sc_wo_ground_truth_df shape: {cell_wout_ground_truth_df.shape[0]}")
+
+
+# ### Fetch the indices from each ground truth and data split and add the status back to sc_profile
+
+# In[11]:
+
+
+# flatten each list in the dictionary
+import itertools
```
Comment (gwaybio) [inline 7.sc_timelapse_model/scripts/1.data.splitting.py:242]: imports at top of file

File: 7.sc_timelapse_model/scripts/1.data.splitting.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+        "ground_truth": "Metadata_ground_truth_present",
+    },
+    inplace=True,
+)
+
+
+# In[14]:
+
+
+# final breakdown of the data
```
Comment (gwaybio) [inline 7.sc_timelapse_model/scripts/1.data.splitting.py:293]: great to see the assertions at the end of this, but I'm confused - do you use these variables later? I just see writing at the end, but a different file?

**Review by gwaybio — APPROVED** [review]
Logic LGTM, but documentation could use a lot of improvements - at many places I was confused and a quick note could alleviate

File: 7.sc_timelapse_model/scripts/1.data.splitting.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# This is quite the complex data splitting procedure.
+# The data is split into holdout data, training, validation, and testing.
+# The training and validation data only contains single-cells that have grond truth labels at the terminal time point.
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 7.sc_timelapse_model/scripts/1.data.splitting.py:6]: fixed typo

File: 7.sc_timelapse_model/scripts/1.data.splitting.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+# While testing and holdout data contains celles that do and no not have ground truth labels at the terminal time point.
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 7.sc_timelapse_model/scripts/1.data.splitting.py:7]: fixed typo

File: 7.sc_timelapse_model/scripts/0.data_cleaning.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+print(sc_well_fovs_endpoint_df.shape)
+sc_well_fovs_endpoint_df.reset_index(drop=True, inplace=True)
+sc_well_fovs_endpoint_df["Metadata_Time"] = 13.0
+sc_well_fovs_endpoint_df.head()
+
+
+# In[11]:
+
+
+# find the unique values in the Metadata_sc_unique_track_id column that have more than 11 time values
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 7.sc_timelapse_model/scripts/0.data_cleaning.py:168]: removed this comment

File: 7.sc_timelapse_model/scripts/0.data_cleaning.py (hunk 2 above)
Comment (MikeLippincott) [reply 7.sc_timelapse_model/scripts/0.data_cleaning.py:184]: Added documentation here: The reason is that we cannot merge these two profiles as they have drastically different datasets and thus would have a lot of NA values

File: 7.sc_timelapse_model/scripts/1.data.splitting.py [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+
+# At this point there are two subsets to the dataset and will be split into the following datasplits:
+# - Single-cells that have a single cell ground truth
+#      - Holdout wells: 1/3 of wells
+#     - Train: 80% of the all cells with a single cell ground truth from the non-holdout wells
+#     - Validation:  10% of the all cells with a single cell ground truth from the non-holdout wells
+#     - Test: 10% of the all cells with a single cell ground truth from the non-holdout wells
+# - Single-cells that do not have a single cell ground truth
+#     - Holdout wells: 1/3 of wells
+#     - Test: 100% of the all cells with a single cell ground truth from the non-holdout wells
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 7.sc_timelapse_model/scripts/1.data.splitting.py:84]: They will have a two column truth.
the data split column and the ground truth column this will diverge both

File: 7.sc_timelapse_model/scripts/1.data.splitting.py (hunk 4 above)
Comment (MikeLippincott) [reply 7.sc_timelapse_model/scripts/1.data.splitting.py:153]: added!

File: 7.sc_timelapse_model/scripts/1.data.splitting.py (hunk 5 above)
Comment (MikeLippincott) [reply 7.sc_timelapse_model/scripts/1.data.splitting.py:242]: moved

File: 7.sc_timelapse_model/scripts/1.data.splitting.py [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+
+# ### Fetch the indices from each ground truth and data split and add the status back to sc_profile
+
+# In[11]:
+
+
+# flatten each list in the dictionary
+import itertools
+
+for key in index_data_split_and_ground_truth_dict.keys():
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 7.sc_timelapse_model/scripts/1.data.splitting.py:244]: I have two columns to contain both informations

File: 7.sc_timelapse_model/scripts/1.data.splitting.py (hunk 6 above)
Comment (MikeLippincott) [reply 7.sc_timelapse_model/scripts/1.data.splitting.py:293]: I added for comments to explain here:

```
# write the data splits to a parquet file
# this writes the indexes, ground truth, and data splits to a parquet file
# we do not write the sc_profile dataframe to a parquet file
```
The assertions are derived from the indexes so it all works out.
